### ISSUE TITLE
Database Schema for OCPI Implementation with SQL vs NoSQL

### ISSUE THREAD

**Issue by RoTorEx** [ocpi/ocpi#514]
Hi everyone,

I'm currently working on a project that requires implementing basic OCPI functionality. Our team has chosen MongoDB for storing OCPI objects, but I'm concerned this might not be the best approach. The OCPI protocol involves significant object relationships, and I believe a SQL database like PostgreSQL with robust object-relational mapping would be more appropriate.

However, due to my limited understanding of the protocol, I’m struggling to design a consistent and efficient table model. It seems logical that if the OCPI protocol is standardized, there should be a recommended or commonly used database schema to ensure optimal implementation.

I’m reaching out to the community to ask for guidance. If you have experience with OCPI in production, could you share insights or examples of effective database schemas (including object relationships)? Your help would be immensely appreciated!

Thanks in advance! :)

**Comment by RoTorEx** [comment]
Upd.

So, here's what we have so far in MongoDB, and it looks awful and I really what to see strict SQL schema instead of this...

```
{
  "party_name": "string", 
  "token_a": "string_or_null",
  "token_b": "string_or_null",
  "token_c": "string_or_null",
  "available_versions_url": "string_or_null",
  "roles": [
    {
      "role": "string",
      "party_id": "string",
      "country_code": "string",
      "business_details": {
        "name": "string",
        "website": "string_or_null"
      }
    }
  ],
  "version": "2.2.1",
  "kind": "sender_or_receiver_enum",
  "status": "kind_of_obj_status_enum",
  "created_at": "datetime"
}
```

**Comment by Sealos** [comment]
As you correctly pointed out, OCPI relies heavily on relationships, and modeling these in NoSQL is non-trivial. 

I would **highly** suggest using SQL. 

TLDR: The OCPI spec is almost a SQL database schema. Don't reinvent the wheel.

**Comment by RoTorEx** [comment]
> As you correctly pointed out, OCPI relies heavily on relationships, and modeling these in NoSQL is non-trivial.
> 
> I would **highly** suggest using SQL.
> 
> TLDR: The OCPI spec is almost a SQL database schema. Don't reinvent the wheel.

Hi @Sealos,

Many thanks for your reply. You’re absolutely right – SQL is much better suited for implementing OCPI due to the complex object relationships as I see.

I’m a bit confused as to why there isn’t an official or at least recommended database schema for an open and widely accepted protocol like OCPI, which clearly defines all interactions and object relationships. This lack of a schema makes designing a database structure quite challenging.

At the moment, I’m working with OCPI for the first time (I'm the only one backend dev here), and our company is only utilizing a limited number of modules from the protocol. However, we plan to expand functionality in the future, which raises serious concerns. If the database schema is poorly designed now, it could lead to scalability issues or even require a complete overhaul later.

If you have an example of a SQL schema that covers all or at least the key aspects of OCPI, I would be incredibly grateful if you could share it. This would help us avoid mistakes and save a lot of time.

I’m even willing to consider paying for your help if it requires significant effort on your part.

I'm looking forward for your response!

**Comment by BazJirIpekyuz** [comment]
Although I haven't implemented the OCPI protocol yet, I analyzed the OCPI endpoint responses and CDRs some months ago, and found a lot of inner relationships in them, so in my opinion, implementing with sql could be quite complex,and querying data could be a big problem with sql, which would require a lot of joins. Therefore, it could really slow down query performance quite a lot.

Because of the a lot of inner objects in CDRs, it seems to me MongoDB is righter option.

**Comment by Sealos** [comment]
> > As you correctly pointed out, OCPI relies heavily on relationships, and modeling these in NoSQL is non-trivial.
> 
> > 
> 
> > I would **highly** suggest using SQL.
> 
> > 
> 
> > TLDR: The OCPI spec is almost a SQL database schema. Don't reinvent the wheel.
> 
> 
> 
> Hi @Sealos,
> 
> 
> 
> Many thanks for your reply. You’re absolutely right – SQL is much better suited for implementing OCPI due to the complex object relationships as I see.
> 
> 
> 
> I’m a bit confused as to why there isn’t an official or at least recommended database schema for an open and widely accepted protocol like OCPI, which clearly defines all interactions and object relationships. This lack of a schema makes designing a database structure quite challenging.
> 
> 
> 
> At the moment, I’m working with OCPI for the first time (I'm the only one backend dev here), and our company is only utilizing a limited number of modules from the protocol. However, we plan to expand functionality in the future, which raises serious concerns. If the database schema is poorly designed now, it could lead to scalability issues or even require a complete overhaul later.
> 
> 
> 
> If you have an example of a SQL schema that covers all or at least the key aspects of OCPI, I would be incredibly grateful if you could share it. This would help us avoid mistakes and save a lot of time.
> 
> 
> 
> I’m even willing to consider paying for your help if it requires significant effort on your part.
> 
> 
> 
> I'm looking forward for your response! 

OCPI defines a standard for communication and relationships of the objects, if you want examples of implementation you will find plentiful by googling a bit.

As I mentioned, you will almost find a database schema (what you are asking) in the OCPI documentation. I wager reading it, or using a tool like ChatGPT would get you further than asking for others for their OCPI implementation.

**Comment by RoTorEx** [comment]
> Although I haven't implemented the OCPI protocol yet, I analyzed the OCPI endpoint responses and CDRs some months ago, and found a lot of inner objects in each data model.
> 
> In my opinion, implementing with sql would be quite complex,and querying data would be a big problem with sql, which would require a lot of joins. Therefore, it could really slow down query performance quite a lot.
> 
> Because of a lot of inner and nested objects in CDRs and data models, it seems to me MongoDB is righter option.

Yes, indeed. At first view, using MongoDB might seem like a good idea since it allows for more convenient handling of nested entities. However, working with and maintaining related entities across different locations can quickly become a nightmare, especially as the product scales. This is one of my primary concerns.

After my initial exploration of the OCPI documentation, I was also overwhelmed by the complex structures and the multitude of possible connections. Given this, a well-organized SQL database schema, which is better suited for managing linked data, seems crucial for ensuring the protocol is properly implemented and maintained during development.

That said, I’m relatively new to OCPI protocol development, and I might not yet see the full picture. I’d like to call on experienced members of the community to discuss potential pitfalls and challenges that may arise during the implementation of the protocol

**Comment by RoTorEx** [comment]
> OCPI defines a standard for communication and relationships of the objects, if you want examples of implementation you will find plenty by googling a bit.
> 
> As I mentioned, you will almost find a database schema (what you are asking) in the OCPI documentation. I wager by reading it, or using a tool like ChatGPT would get you further than asking for others for their OCPI implementation.
> 
> If you have specific questions about the specs, happy to help with those!

You’re probably referring to figure 24 on page 57 of the OCPI v2.2.1 documentation, which is indeed the only explicit diagram showing object relationships. However, the relationships it outlines are already fairly obvious from the written descriptions.

For example, I’m currently struggling to determine the correct table structure for the Credentials module. It’s unclear where to store multiple tokens, how to update or replace them, and how to link them to an integration table. Below, I’ve included a very basic and admittedly clumsy solution I sketched out quickly, but I’m not confident it’s the best approach.

Deep Googling hasn’t yielded any meaningful results on OCPI database schemas, and that’s exactly why I decided to create this ticket here on GitHub—to seek insights from others who might have already dealt with these challenges. If you have specific resources in mind, I’d greatly appreciate it if you could share them.

Regarding ChatGPT, while it’s helpful initially, it tends to hallucinate when faced with complex specifications. Unfortunately, even a paid model doesn’t seem capable of handling the full depth of OCPI’s context effectively

**Comment by RoTorEx** [comment]
As far as I can see, even you guys are talking about different implementation approaches. And this could be a problem, this strikes me as a significant gap in the protocol’s guidance.

For example, Vourity claims to have implemented the OCPI protocol, but based on thier endpoints and servers behaviour, it seems their implementation struggles in practice. And I had to rely on trial-and-error methods to understand their servers' returns, as the behavior of their system doesn’t align with the protocol specifications.

**Comment by julius-yang** [comment]
For us (an eMSP), each top level OCPI class is a table in a relational db -- Location, EVSE, Connector, Tariffs, CDRs, and so on.  

> For example, I’m currently struggling to determine the correct table structure for the Credentials module. It’s unclear where to store multiple tokens, how to update or replace them, and how to link them to an integration table. Below, I’ve included a very basic and admittedly clumsy solution I sketched out quickly, but I’m not confident it’s the best approach.

Instead of a table for a Credentials class we define a CPO table.  A CPO is not an official class in OCPI, but a CPO table where each row is a CPO and columns define Credentials attributes is more convenient for us.   For example the CPO table can have a "name" column instead of nesting it within Credentials.roles[0].business_details.  This also allows us to define non-OCPI columns such as "enabled" (a killswitch for CPOs who are sending excessive OCPI PUTs or PATCHes) or whether the CPO requires pagination.

Our EVSE table has a foreign key to Location.  A Connector has a foreign key to EVSE.  We have intermediate tables for many-to-many relationships such as Connector and Tariff -- one Connector can have multiple Tariffs, and one Tariff can be assigned to multiple Connectors.

Here is the most important decision we made.  **We store many of the smaller sub-classes like Location.operator as a JSON column instead of creating a table for every sub-class.**  This avoids a nightmare tangle of foreign key relationships.  It leads to data redundancy but our primary goal was an understandable schema.  We use an ORM which converts the relationships and JSON columns into a proper OCPI class (a JSON object), and vice versa.

For obvious reasons I can't give you a schema, but I hope this helps.
